### PR TITLE
imp(json-schema): memoize hashCode and make schemaMap immutable

### PR DESCRIPTION
**Pull request by michaelcarter-wf**
## Ultimate problem:
- Performance on `JsonSchema.resolveRef` got noticeably worse since 3.2.1 because of our naive implementation of `hashCode` / `==`. 

## How it was fixed:
- Since we don't allow changes to `schemaMap` / `schemaBool` post-construction, make them `final` / `unmodifiable`.
- Since `hashCode` is based on `schemaMap` / `schemaBool`, we can cache it's value
- Since `==` is also derived from `schemaMap` / `schemaBool`, and `hashCode` is derived from those, use `hashCode` to check equality.

## Potential areas of regression:

---

> __FYA:__ @michaelcarter-wf

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/json_schema/json_schema.dart [hunk 1]
```diff
@@ -124,7 +125,7 @@ class JsonSchema {
     List<CustomVocabulary> customVocabularies,
     Map<String, Map<String, SchemaPropertySetter>> customVocabMap,
     Map<String, ValidationContext Function(ValidationContext context, String instanceData)> customFormats = const {},
-  }) {
+  }) : _schemaMap = null {
```
Comment (robbielamb-wf) [inline lib/src/json_schema/json_schema.dart:128]: I had forgotten about this initialization trick in dart.

Comment (joshriesenbach-wk) [reply lib/src/json_schema/json_schema.dart:128]: Right?! I had to ping Michael to be like "What the heck is that?"

**Review by joshriesenbach-wk — DISMISSED** [review]
+1

**Review by robbielamb-wf — DISMISSED** [review]

**Review by robbielamb-wf — APPROVED** [review]

**Review by robbielamb-wf — APPROVED** [review]

**Comment by michaelcarter-wf** [comment]
+10

- [x] CI Passes
- [x] Solves performance issue with consuming repo

**Review by rmconsole-wf — APPROVED** [review]
+1 from RM

**Pull request merged**